File: Android/app/src/main/java/catglo/com/deliverydroid/neworder/TimeFragment.java
package catglo.com.deliverydroid.neworder;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.NumberPicker;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.TimePicker;

import catglo.com.deliveryDatabase.Order;
import catglo.com.deliverydroid.R;
import catglo.com.deliverydroid.Tools;


import org.joda.time.DateTime;
import org.joda.time.DurationFieldType;
import org.joda.time.Hours;
import org.joda.time.Minutes;
import org.joda.time.MutableDateTime;

/**
 * Created by goblets on 2/17/14.
 */
public class TimeFragment extends DataAwareFragment implements TimePicker.OnTimeChangedListener, NumberPicker.OnValueChangeListener {

    private SharedPreferences sharedPreferences;
    private NumberPicker inHowManyMinutes;
    private TimePicker timePicker;
    private View button1;
    private View button2;
    private View button3;
    private View nextButton;
    private MutableDateTime time;
    private boolean isInit=false;
    private TextView timeLabel;
    private TextView inMinutes;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

    }

    @Override
    public void onResume(){
        super.onResume();

        time = new MutableDateTime(((NewOrderActivity)getActivity()).order.time.getTime());

        timeLabel.setText(Tools.getFormattedTimeDay(time.toGregorianCalendar()));

        nextButton.setOnClickListener(new View.OnClickListener() { public void onClick(final View v) {
            ((ButtonPadFragment.ButtonPadNextListener)getActivity()).onNextButtonPressed();
        }});

        timePicker.setOnTimeChangedListener(this);
        inHowManyMinutes.setOnValueChangedListener(this);

        setButtonClickListenerForMinutes(15,button1);
        setButtonClickListenerForMinutes(25,button2);
        setButtonClickListenerForMinutes(35,button3);

        setDatePicker();
        setNumberPicker();

        InputMethodManager imm = (InputMethodManager)getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        if(imm != null){
            imm.toggleSoftInput(0, InputMethodManager.HIDE_IMPLICIT_ONLY);
        }
    }

    void setButtonClickListenerForMinutes(final int minutes,View button){
        button.setOnClickListener(new View.OnClickListener() {public void onClick(View v) {
            time = MutableDateTime.now();
            time.add(Minutes.minutes(minutes));
            NewOrderActivity activity = ((NewOrderActivity) getActivity());
            Order order = activity.order;
            order.time.setTime(time.getMillis());
            setDatePicker();
            setNumberPicker();
        }});
    }

    public void onPause(){
        super.onPause();  setDatePicker();
        setNumberPicker();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity().getApplicationContext());

        View view = (RelativeLayout) inflater.inflate(R.layout.new_order_time_frgament,null);

        inHowManyMinutes = (NumberPicker)view.findViewById(R.id.numberPicker);
        timePicker = (TimePicker)view.findViewById(R.id.timePicker);

        button1 = view.findViewById(R.id.button1);
        button2 = view.findViewById(R.id.button2);
        button3 = view.findViewById(R.id.button3);

        nextButton = view.findViewById(R.id.next_button);

        inHowManyMinutes.setMaxValue(60);
        inHowManyMinutes.setMinValue(1);

        timeLabel = (TextView)view.findViewById(R.id.timeLabel);
        inMinutes = (TextView)view.findViewById(R.id.inMinutes);
        return view;
    }

    @Override
    protected void onDataChangedHandler() {
        time = new MutableDateTime(((NewOrderActivity)getActivity()).order.time.getTime());
        setDatePicker();
        setNumberPicker();
    }

    public void setDatePicker(){
        timePicker.setCurrentHour(time.getHourOfDay());
        timePicker.setCurrentMinute(time.getMinuteOfHour());
    }

    public void setNumberPicker(){
        DateTime now = DateTime.now();
        int minutesToOrder = Minutes.minutesBetween(now, time).getMinutes();
        if (minutesToOrder<0 || minutesToOrder>59){
            inHowManyMinutes.setVisibility(View.INVISIBLE);
        } else {
            inHowManyMinutes.setVisibility(View.VISIBLE);
            inHowManyMinutes.setValue(minutesToOrder);
        }
    }


    @Override
    public void onTimeChanged(TimePicker view, int hourOfDay, int minute) {
        if (isInit==false){
            NewOrderActivity activity = ((NewOrderActivity) getActivity());
            Order order = activity.order;
            time.setHourOfDay(hourOfDay);
            time.setMinuteOfHour(minute);

            MutableDateTime thirtyMinutesAgo = MutableDateTime.now();
            thirtyMinutesAgo.add(Minutes.minutes(-30));
            if (time.isBefore(thirtyMinutesAgo)){
                time.add(Hours.hours(24));
            }

            MutableDateTime twentyFourHoursFromNow = MutableDateTime.now();
            twentyFourHoursFromNow.add(Hours.hours(24));
            if (time.isAfter(twentyFourHoursFromNow)){
                time.add(Hours.hours(-24));
            }


            order.time.setTime(time.getMillis());
            isInit=true;
            setNumberPicker();
            isInit=false;
            DataAwareFragment lastScreenFragment = (DataAwareFragment)  activity.getFragment(NewOrderActivity.Pages.order);
            if (lastScreenFragment!=null) lastScreenFragment.onDataChanged();
        }
        timeLabel.setText(Tools.getFormattedTimeDay(time.toGregorianCalendar()));
        inMinutes.setText("In "+Minutes.minutesBetween(DateTime.now(),time).getMinutes()+" minutes");

    }

    @Override
    public void onValueChange(NumberPicker picker, int oldVal, int newVal) {
        if (isInit==false){
            NewOrderActivity activity = ((NewOrderActivity) getActivity());
            Order order = activity.order;
            time = new MutableDateTime(DateTime.now());
            time.add(DurationFieldType.minutes(),newVal);
            order.time.setTime(time.getMillis());
            isInit=true;
            setDatePicker();
            isInit=false;
            DataAwareFragment lastScreenFragment = (DataAwareFragment)  activity.getFragment(NewOrderActivity.Pages.order);
            if (lastScreenFragment!=null) lastScreenFragment.onDataChanged();
        }
    }
}

File: Android/app/src/main/java/catglo/com/deliverydroid/neworder/TimeFragment.kt
package catglo.com.deliverydroid.neworder

import android.content.SharedPreferences
import android.content.res.Resources
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.NumberPicker
import android.widget.TextView
import android.widget.TimePicker

import catglo.com.deliverydroid.R
import catglo.com.deliverydroid.widgets.HorizontalDatePicker
import kotlinx.android.synthetic.main.new_order_time_frgament.*


import org.joda.time.DateTime
import org.joda.time.DurationFieldType
import org.joda.time.Minutes
import org.joda.time.MutableDateTime
import kotlin.math.absoluteValue
import android.app.Activity
import android.util.Log
import android.view.View.*


/**
 * Created by goblets on 2/17/14.
 */
class TimeFragment : DataAwareFragment(), TimePicker.OnTimeChangedListener, NumberPicker.OnValueChangeListener {

    private var sharedPreferences: SharedPreferences? = null
    private var inHowManyMinutes: NumberPicker? = null
    private var timePicker: TimePicker? = null
    private var nextButton: View? = null
    private var time = MutableDateTime(MutableDateTime.now())
    private var isInit = false
    private var datePicker: HorizontalDatePicker? = null
    private var inMinutes: TextView? = null

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

    }

    override fun onResume() {
        super.onResume()

        (activity as NewOrderActivity).tools.hideOnScreenKeyboard()
        datePicker?.setTime(time)

        nextButton?.setOnClickListener { (activity as ButtonPadFragment.ButtonPadNextListener).onNextButtonPressed() }

        timePicker?.setOnTimeChangedListener(this)
        inHowManyMinutes?.setOnValueChangedListener(this)


        setDatePicker()
        setNumberPicker()

        val imm = activity?.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        //Find the currently focused view, so we can grab the correct window token from it.
        var view = activity?.getCurrentFocus()
        //If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) {
            view = View(activity)
        }
        imm.hideSoftInputFromWindow(view.windowToken, 0)

        timePicker?.findViewById<View>(Resources.getSystem().getIdentifier("minutes","id", "android"))?.let {
            it.post {  it.performClick() }
        }
        timePicker?.findViewById<View>(Resources.getSystem().getIdentifier("toggle_mode","id", "android"))?.let {
            it.post {  it.visibility = GONE }
        }
        datePicker?.onDaySelected = onDayChanged()


    }


    override fun onPause() {
        super.onPause()
        setDatePicker()
        setNumberPicker()
    }

    var once = true
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity?.applicationContext)

        val view = inflater.inflate(R.layout.new_order_time_frgament, null)

        inHowManyMinutes = view.findViewById<View>(R.id.numberPicker) as NumberPicker
        timePicker = view.findViewById<View>(R.id.timePicker) as TimePicker

        nextButton = view.findViewById(R.id.next_button)

        inHowManyMinutes?.maxValue = 181
        inHowManyMinutes?.minValue = 0
        inHowManyMinutes?.setFormatter {
            if (it<=91) it.toString()
            else (0-(182-it)).toString()
        }
        datePicker = view.findViewById(R.id.timeLabel)
        inMinutes = view.findViewById<View>(R.id.inMinutes) as TextView

        val nowButton = view.findViewById<View>(R.id.nowButton)
        nowButton.setOnClickListener {
            updateAllPickers( MutableDateTime(MutableDateTime.now()))
        }

        return view
    }

    fun numberPickerFirstNumberHack(){
        try {
            val method = numberPicker!!.javaClass.getDeclaredMethod("changeValueByOne", Boolean::class.javaPrimitiveType)
            method.setAccessible(true)
            method.invoke(numberPicker!!, true)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }


    fun updateAllPickers(t:MutableDateTime){
        val activity = activity as NewOrderActivity? ?: return
        if (isInit) return
        isInit = true
        Log.i("DATE","updateAllPickers ${t.toString()} ")
        time = t
        datePicker?.setTime(time)
        setDatePicker()
        setNumberPicker()
        val timeDiff = time.dayOfYear - MutableDateTime.now().dayOfYear().get()
        inHowManyMinutes?.visibility = if (timeDiff == 0) { VISIBLE } else { GONE }
        val lastScreenFragment = activity.getFragment(NewOrderActivity.Pages.order)
        lastScreenFragment?.onDataChanged()
        isInit = false
    }

    override fun onDataChangedHandler() {
        time = MutableDateTime((activity as NewOrderActivity).order.time.time)
        updateAllPickers(time)
    }

    fun setDatePicker() {
        time.let { t->
            timePicker?.currentHour = t.hourOfDay
            timePicker?.currentMinute = t.minuteOfHour
        }
    }


    var blockSetNumberPicker = false

    fun setNumberPicker() {
        if (blockSetNumberPicker) return
        val now = DateTime.now()
        time?.let {t ->
            val minutesToOrder = Minutes.minutesBetween(now, t).minutes
            if (minutesToOrder in 0..91){
                inHowManyMinutes?.value = minutesToOrder
                inHowManyMinutes?.visibility = VISIBLE
                numberPickerFirstNumberHack()
            }
            else if (minutesToOrder in -90..0){
                val value = 182+minutesToOrder
                inHowManyMinutes?.value = value
                inHowManyMinutes?.visibility = VISIBLE
                numberPickerFirstNumberHack()
            }
            else {
                inHowManyMinutes?.visibility = GONE
            }
        }
    }


    /***************
     * Callbacks for each of the 3 controls
     */
    override fun onTimeChanged(view: TimePicker, hourOfDay: Int, minute: Int) {
        val activity = activity as NewOrderActivity?

        Log.i("DATE","onTimeChanged $hourOfDay : $minute ")

        val order = activity?.order
        time?.hourOfDay = hourOfDay
        time?.minuteOfHour = minute
        time?.let { t ->
            order?.time?.time = t.millis
        }
        if (isInit == false) {

            isInit = true
            updateAllPickers(time)
            isInit = false

        }
        time?.let { t ->
            datePicker?.setTime(t)
            val minutesDifference =  Minutes.minutesBetween(DateTime.now(), t).minutes
            inMinutes?.text = if (minutesDifference>0) "In $minutesDifference minutes" else "${-minutesDifference} minutes ago"
        }
    }

    override fun onValueChange(picker: NumberPicker, oldVal: Int, newVal: Int) {
        if (isInit == false) {
            Log.i("DATE","onValueChange $newVal ")
            blockSetNumberPicker = true
            val activity = activity as NewOrderActivity?
            val order = activity?.order
            time = MutableDateTime(DateTime.now())
            time?.let { t ->
                val value = if (newVal<=91) newVal else (0-(182-newVal))
                t.add(DurationFieldType.minutes(), value)
                order?.time?.time = t.millis
                val lastScreenFragment = activity?.getFragment(NewOrderActivity.Pages.order)
                timePicker?.currentHour = t.hourOfDay.absoluteValue
                timePicker?.currentMinute = t.minuteOfHour.absoluteValue
                time = t
                updateAllPickers(t)
            }
            blockSetNumberPicker = false
        }
    }

    private fun onDayChanged(): (Int) -> Unit {
        return {
            Log.i("DATE","onDayChanged $it ")
            time.dayOfYear = MutableDateTime.now().dayOfYear().get() + it
            updateAllPickers(time)
        }
    }

}
